Add httpstatus 415, unsupported media type

bf4 opened the pull request:
- [x] Add ActionController::UnsupportedMediaType
- [x] explain difference from `UnknownFormat` which is a 'bad_request' (400)
  - per @rafaelfranca 
    - request `Accept` header which matches no known media types should raise `UnknownFormat` (i.e. we don't know understand any types the acceptable to respond to the client)
   - request `Content-Type` header which isn't a known media type should raise `UnsupportedMediaType` (i.e. we don't have a registered media type that understands the request's content type)
  - (look at  config.action_dispatch.rescue_responses )
- [x] Add failing test
- [x] raise failure at appropriate point in mime negotiation

Per
- discussion in https://github.com/rails/rails/issues/26309#issuecomment-249673783


https://tools.ietf.org/html/rfc7231#section-6.5.13 (via https://httpstatuses.com/415 )

>  The 415 (Unsupported Media Type) status code indicates that the
   origin server is refusing to service the request because the payload
   is in a format not supported by this method on the target resource.
   The format problem might be due to the request's indicated
   Content-Type or Content-Encoding, or as a result of inspecting the
   data directly.

bf4 committed:
Add httpstatus 415, unsupported media type

bf4 committed:
Failing test

rafaelfranca commented on actionpack/test/dispatch/request/json_params_parsing_test.rb:
```diff
@@ -38,10 +38,13 @@ def teardown
-      {},
-      "{\"person\": {\"name\": \"David\"}}", "CONTENT_TYPE" => "application/vnd.api+json"
-    )
+    actual = "{\"person\": {\"name\": \"David\"}}", "CONTENT_TYPE" => "application/vnd.api+json"
```
This seems wrong. Should not `"CONTENT_TYPE" => "application/vnd.api+json"` be the headers?

bf4 replied:
well, that's embarrassing... syntax error

bf4 replied:
(in case not clear, I've corrected the error, and followed up with a commit that adds comments in possible touch points)

rafaelfranca reviewed (changes requested):
I like the idea. Let's move it forward.

bf4 committed:
Where should we raise the error and where do we rescue it?

bf4 commented on actionpack/test/dispatch/request/json_params_parsing_test.rb:
```diff
@@ -38,10 +38,13 @@ def teardown
+    assert_raise(ActionController::UnsupportedMediaType) {
+      with_test_routing do
+        post "/parse", params: actual, headers: headers
+      end
```
Right now this, besides not raising an exception, returns an empty hash `{}`, and I haven't yet found where that's coming from.

This test may actually be more appropriate in a mime-responds / mime negotiation test file since it is not really about params parsing, at least in the form in this PR.

bf4 committed:
415

bf4 committed:
debug exceptions

rafaelfranca commented on actionpack/lib/action_controller/metal/mime_responds.rb:
```diff
@@ -200,6 +200,7 @@ def respond_to(*mimes)
         response = collector.response
         response.call if response
       else
+        # PRNOTE(BF): Should this be a 415, UnsupportedMediaType?
```
As it is based in the Accept header `UnknownFormat` is correct here.

rafaelfranca commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -16,6 +16,8 @@ module MimeNegotiation
       # X-Post-Data-Format HTTP header if present.
       def content_mime_type
         fetch_header("action_dispatch.request.content_type") do |k|
+          # PRNOTE(BF): This is where the lookup that triggers the 415 should happen, I think.
```
:+1:

rafaelfranca reviewed (commented):
I think as rule of thumb When dealign with the `Accept` header we should raise `UnknownFormat` and when dealing with the Content-Type header we should raise `UnsupportedMediaType`

bf4 commented:
Thanks for comments, will push changes when back from holiday in a week or so. 🚗🎣🌊🐠🏊🏻

B mobile phone

> On Jan 3, 2017, at 8:34 PM, Rafael França <notifications@github.com> wrote:
> 
> @rafaelfranca commented on this pull request.
> 
> I think as rule of thumb When dealign with the Accept header we should raise UnknownFormat and when dealing with the Content-Type header we should raise UnsupportedMediaType
> 
> In actionpack/lib/action_controller/metal/mime_responds.rb:
> 
> > @@ -200,6 +200,7 @@ def respond_to(*mimes)
>          response = collector.response
>          response.call if response
>        else
> +        # PRNOTE(BF): Should this be a 415, UnsupportedMediaType?
> As it is based in the Accept header UnknownFormat is correct here.
> 
> In actionpack/lib/action_dispatch/http/mime_negotiation.rb:
> 
> > @@ -16,6 +16,8 @@ module MimeNegotiation
>        # X-Post-Data-Format HTTP header if present.
>        def content_mime_type
>          fetch_header("action_dispatch.request.content_type") do |k|
> +          # PRNOTE(BF): This is where the lookup that triggers the 415 should happen, I think.
> 👍
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bf4 committed:
Remove exploratory PR and choose exception origin per rafaelfranca

bf4 committed:
Show exceptions test

bf4 committed:
Fix style

bf4 committed:
Conditionally validate content mime time

Fixes:

```
RequestFormData#test_media_type_is_not_from_the_FORM_DATA_MEDIA_TYPES_array:
ActionController::UnsupportedMediaType: multipart/related
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:25:in `block (2 levels) in content_mime_type'
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:20:in `tap'
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:20:in `block in content_mime_type'
    /home/travis/build/rails/rails/vendor/bundle/ruby/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:57:in `fetch'
    /home/travis/build/rails/rails/vendor/bundle/ruby/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:57:in `fetch_header'
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:18:in `content_mime_type'
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/request.rb:232:in `media_type'
    /home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/request.rb:317:in `form_data?'
    /home/travis/build/rails/rails/actionpack/test/dispatch/request_test.rb:1280:in `block in <class:RequestFormData>'
```

bf4 committed:
Update error assertion on mutating headers

Fixes:

```
TestCaseTest#test_mutating_content_type_headers_for_non_registered_mime_type_raises_an_error [/home/travis/build/rails/rails/actionpack/test/controller/test_case_test.rb:584]:
[RuntimeError] exception expected, not
Class: <ActionController::UnsupportedMediaType>
Message: <"type/fake">
---Backtrace---
/home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:25:in `block (2 levels) in content_mime_type'
/home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:20:in `tap'
/home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:20:in `block in content_mime_type'
/home/travis/build/rails/rails/vendor/bundle/ruby/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:57:in `fetch'
/home/travis/build/rails/rails/vendor/bundle/ruby/2.2.0/gems/rack-2.0.1/lib/rack/request.rb:57:in `fetch_header'
/home/travis/build/rails/rails/actionpack/lib/action_dispatch/http/mime_negotiation.rb:18:in `content_mime_type'
/home/travis/build/rails/rails/actionpack/lib/action_controller/test_case.rb:101:in `assign_parameters'
/home/travis/build/rails/rails/actionpack/lib/action_controller/test_case.rb:495:in `process'
/home/travis/build/rails/rails/actionpack/lib/action_controller/test_case.rb:400:in `post'
/home/travis/build/rails/rails/actionpack/test/controller/test_case_test.rb:586:in `block in test_mutating_content_type_headers_for_non_registered_mime_type_raises_an_error'
---------------
```

bf4 commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -14,10 +14,17 @@ module MimeNegotiation
+              # Otherwise, it returns <tt>Mime::Type.new(string)</tt> which has a the symbol +nil+.
+              next unless content_mime_type.symbol.nil?
+              next if env.key?("action_dispatch.exception") # Don't raise in exceptions_app, else FAILSAFE_RESPONSE
+              raise ActionController::UnsupportedMediaType.new(content_mime_type)
```
@rafaelfranca I implemented raising the 415 as requested when getting the `content_mime_type`, but since there's a test that checks for a form post media type that doesn't expect an exception, I added the `validate: true` default.  I don't really like it... needs thought 86c7c391fce919ac6a547887174c7257d43ff002

bf4 commented on actionpack/test/dispatch/request/json_params_parsing_test.rb:
```diff
@@ -60,14 +65,15 @@ def teardown
+    )
+    output.rewind && err = output.read
+    assert err =~ /Error occurred while parsing request parameters/
+    assert_equal("400 error fixture\n", @response.body)
```
No need to assert response.body here.  I'm ok with removing it.

bf4 commented on actionpack/test/dispatch/request/json_params_parsing_test.rb:
```diff
@@ -38,10 +38,15 @@ def teardown
     )
+    output.rewind && err = output.read
+    assert err =~ /ActionController::UnsupportedMediaType \(#{Regexp.escape("application/vnd.api+json")}\):/
+    assert_equal("", @response.body)
```
No need to assert response.body here.  I'm ok with removing it. It appears to be tested in actionpack/test/dispatch/show_exceptions_test.rb

bf4 commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -229,7 +229,7 @@ def original_url
     #    request.media_type # => "application/x-www-form-urlencoded"
     def media_type
-      content_mime_type.to_s
+      content_mime_type(validate: false).to_s
```
see https://github.com/rails/rails/pull/26632/files#r99506410

bf4 commented:
Ready for review

bf4 commented:
Looks like Basecamp API V3 uses 415 https://github.com/basecamp/bc3-api/blame/161837a102c6da1548f3009e308b967a484d5385/README.md#L86 `¯\_(ツ)_/¯`

bf4 commented:
ref https://github.com/rails/rails/issues/20666 https://github.com/rails/rails/pull/23827 https://github.com/rails/rails/issues/22853

oniofchaos commented:
@bf4 realize this is old but are you still interested in getting this in. Looks like there's some merge conflicts now (understandably). 💞

bf4 commented:
@oniofchaos If there's a chance it'll be looked at by someone with commit.

Yorkshireman commented:
Does anyone have plans to progress this?

bf4 commented:
@y-yagi If there's interest to return to this on rails-core, I'm happy to revive it. However, I've learned that it's not worth my effort without someone on rails-core interested in it.

rafaelfranca commented:
I'm still interested on this. Quick feedback, the current behavior should stay the same, but we should add an option that make Rails validate the media type.

bf4 commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -14,10 +14,17 @@ module MimeNegotiation
       # For backward compatibility, the post \format is extracted from the
       # X-Post-Data-Format HTTP header if present.
-      def content_mime_type
+      def content_mime_type(validate: true)
```
@rafaelfranca seems to me the changeset here could be something like

```ruby
- def content_mime_type
+ def content_mime_type(raise_errors_on_unsupported_media_type: Rails.configuration.raise_errors_on_unsupported_media_type)
```

where `Rails.configuration.raise_errors_on_unsupported_media_type` defaults to `false`

rafaelfranca replied:
Yes, something similar to that. But we can't make Action Pack depend on Rails so we need to inverse this dependency. The Railtie configure Action Pack but the config is on this module.

antropova commented:
has this been scraped altogether?

bf4 commented:
I'm personally done with these sorts of Rails PRs. Ie anything that requires more than one round of feedback always dies on the vine
